File: iOS/AmazonConnectInAppCallingIOSSample/AmazonConnectInAppCallingIOSSample/InAppCallingComponent/Services/Api/ApiClient.swift
//
//  ApiClient.swift
//
//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
//  SPDX-License-Identifier: MIT-0
//

import Foundation

// TODO: remove
struct CreateMeetingResponse: Codable {

    let joinInfo: JoinInfoData
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        joinInfo = try container.decode(JoinInfoData.self, forKey: .joinInfo)
    }
    
    enum CodingKeys: String, CodingKey {
        case joinInfo = "JoinInfo"
    }
}

// TODO: remove
struct JoinInfoData: Codable {
    
    let attendee: AttendeeData
    let meeting: MeetingData
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        attendee = try container.decode(AttendeeData.self, forKey: .attendee)
        meeting = try container.decode(MeetingData.self, forKey: .meeting)
        
    }
    
    enum CodingKeys: String, CodingKey {
        case meeting = "Meeting"
        case attendee = "Attendee"
    }
}

// TODO: remove
struct JoinInfoMeetingData: Codable {
    
    let meeting: MeetingData
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        meeting = try container.decode(MeetingData.self, forKey: .meeting)
        
    }
    
    enum CodingKeys: String, CodingKey {
        case meeting = "Meeting"
    }
}

// TODO: remove
struct JoinInfoAttendeeData: Codable {
    
    let attendee: AttendeeData
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        attendee = try container.decode(AttendeeData.self, forKey: .attendee)
        
    }
    
    enum CodingKeys: String, CodingKey {
        case attendee = "Attendee"
    }
}

class ApiClient {
    
    private let config: InAppCallingConfiguration
    private let httpClient: HttpClient
    
    init(config: InAppCallingConfiguration,
         httpClient: HttpClient) {
        self.config = config
        self.httpClient = httpClient
    }
    
    
    // TODO: remove
    func createWebrtcContact(connectInstanceId: String,
                             contactFlowId: String,
                             displayName: String,
                             attributes: [String: String],
                             onSuccess: @escaping (_ response: CreateWebrtcContactResponse) -> Void,
                             onFailure: @escaping (_ error: Error) -> Void) {
        let url = "https://4w5z5lp6f3.execute-api.us-east-1.amazonaws.com/Prod/join"
        let payload = [
            "attendeeName": "george",
            "title": "screenshare16",
            "region": "us-west-2"
        ]
        
        self.httpClient.postJson(url, nil, payload) { data in
            let meetingInfo: CreateMeetingResponse = data
            
            let response = CreateWebrtcContactResponse(meetingData: meetingInfo.joinInfo.meeting,
                                                       attendeeData: meetingInfo.joinInfo.attendee,
                                                       contactId: "", participantId: "", participantToken: "")
            onSuccess(response)
        } _: { error in
            onFailure(error)
        }


    }
    
//    func createWebrtcContact(connectInstanceId: String,
//                             contactFlowId: String,
//                             displayName: String,
//                             attributes: [String: String],
//                             onSuccess: @escaping (_ response: CreateWebrtcContactResponse) -> Void,
//                             onFailure: @escaping (_ error: Error) -> Void) {
//        let url = self.config.startWebrtcContactEndpoint
//        let body = CreateWebrtcContactRequest(connectInstanceId: connectInstanceId,
//                                              contactFlowId: contactFlowId,
//                                              displayName: displayName,
//                                              attributes: attributes)
//        self.httpClient.postJson(url, nil, body) { data in
//            onSuccess(data)
//        } _: { error in
//            onFailure(error)
//        }
//    }
    
    func createParticipantConnection(participantToken: String,
                                     onSuccess: @escaping (_ response: CreateParticipantConnectionResponse) -> Void,
                                     onFailure: @escaping (_ error: Error) -> Void) {
        let url = self.config.createParticipantConnectionEndpoint
        let body = ["ParticipantToken": participantToken]
        self.httpClient.postJson(url, nil, body) { data in
            onSuccess(data)
        } _: { error in
            onFailure(error)
        }
    }
    
    func sendMessage(connectionToken: String,
                     digits: String,
                     onSuccess: @escaping () -> Void,
                     onFailure: @escaping (_ error: Error) -> Void) {
        let url = self.config.sendMessageEndpoint
        let body = [
            "ConnectionToken": connectionToken,
            "Digits": digits
        ]
        self.httpClient.postJson(url, nil, body) {
            onSuccess()
        } _: { error in
            onFailure(error)
        }
    }
}

File: Android/AmazonConnectInAppCallingAndroidSample/app/src/main/java/com/amazonaws/services/connect/inappcalling/sample/data/api/ApiClient.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 */

package com.amazonaws.services.connect.inappcalling.sample.data.api

import com.amazonaws.services.chime.sdk.meetings.internal.utils.HttpUtils
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import com.google.gson.Gson
import com.amazonaws.services.connect.inappcalling.sample.data.utils.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URL

data class ApiClientConfig(
    val startWebRTCContactEndpoint: String,
    val createParticipantConnectionEndpoint: String,
    val sendMessageEndpoint: String
)
interface ApiClient {
    suspend fun createWebrtcContact(request: CreateWebrtcContactRequest)
            : Result<CreateWebrtcContactResponse>

    suspend fun createParticipantConnection(participantToken: String)
            : Result<CreateParticipantConnectionResponse>

    suspend fun sendMessage(
        connectionToken: String,
        digits: String
    ): Result<Unit>
}
class DefaultApiClient(private val config: ApiClientConfig) : ApiClient {
    private val logger = ConsoleLogger()
    private val tag = "ApiClient"

    override suspend fun createWebrtcContact(request: CreateWebrtcContactRequest)
    : Result<CreateWebrtcContactResponse>  = withContext(Dispatchers.IO) {
        logger.info(tag, "createWebrtcContact with request: $request")
        val errorMsg = "Failed to createWebrtcContact for request $request, due to: "
        try {
            val url = config.startWebRTCContactEndpoint
            val res = HttpUtils.post(
                url = URL(url),
                body = Gson().toJson(request),
                logger = logger
            )
            logger.info(tag, "response: ${res.data}")
            val response = Gson().fromJson(res.data!!, CreateWebrtcContactResponse::class.java)
            Result.Success(response)
        } catch (e: Exception) {
            logger.error(tag, errorMsg + e.message)
            Result.Failure(Error(errorMsg + e.message))
        }
    }

    override suspend fun createParticipantConnection(participantToken: String)
    : Result<CreateParticipantConnectionResponse> = withContext(Dispatchers.IO) {
        logger.info(tag, "createParticipantConnection with participantToken: $participantToken")
        val errorMsg = "Failed to createParticipantConnection for participantToken $participantToken, due to:"
        try {
            val url = config.createParticipantConnectionEndpoint
            val res = HttpUtils.post(
                url = URL(url),
                body = "{ \"ParticipantToken\": \"$participantToken\"}",
                logger = logger
            )
            logger.info(tag, "response: ${res.data}")
            val response = Gson().fromJson(res.data!!, CreateParticipantConnectionResponse::class.java)
            Result.Success(response)
        } catch (e: Exception) {
            logger.error(tag, errorMsg + e.message)
            Result.Failure(Error(errorMsg + e.message))
        }
    }

    override suspend fun sendMessage(
        connectionToken: String,
        digits: String
    ): Result<Unit>  = withContext(Dispatchers.IO) {
        logger.info(tag, "sendMessage with digits: $digits")
        val errorMsg = "Failed to sendMessage for digits: $digits, connectionToken: $connectionToken, due to: "
        try {
            val url = config.sendMessageEndpoint
            val res = HttpUtils.post(
                url = URL(url),
                body = "{ \"ConnectionToken\": \"$connectionToken\", \"Digits\": \"$digits\"}",
                logger = logger
            )
            logger.info(tag, "response: ${res.data}")
            if (res.httpException == null) {
                Result.Success(Unit)
            } else {
                logger.error(tag, errorMsg + res)
                Result.Failure(Error(errorMsg + res))
            }
        } catch (e: Exception) {
            logger.error(tag, errorMsg + e.message)
            Result.Failure(Error(errorMsg + e.message))
        }
    }
}
